hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java

```diff
@@ -809,6 +882,246 @@ public Bytes getFailureResultFor(final ResponseCodeEnum status) {
+				case ABI_ID_CREATE_NON_FUNGIBLE_TOKEN -> decoder.decodeNonFungibleCreate(input, aliasResolver);
+				case ABI_ID_CREATE_NON_FUNGIBLE_TOKEN_WITH_FEES ->
+						decoder.decodeNonFungibleCreateWithFees(input, aliasResolver);
+				default -> null;
+			};
+
+			/* --- Validate Solidity input and massage it to be able to transform it to tokenCreateTxnBody --- */
+			verifySolidityInput();
+			try {
+				replaceInheritedKeysWithSenderKey();
+			} catch (DecoderException e) {
+				throw new InvalidTransactionException(FAIL_INVALID);
+			}
+
+			return syntheticTxnFactory.createTokenCreate(tokenCreateOp);
+		}
+
+		@Override
+		public void run(final MessageFrame frame)  {
+			Objects.requireNonNull(tokenCreateOp);
+
+			/* --- Validate the synthetic create txn body before proceeding with the rest of the execution --- */
+			final var creationTime = recordsHistorian.nextFollowingChildConsensusTime();
+			final var result = tokenCreateChecks.validatorForConsTime(creationTime).apply(transactionBody.build());
+			validateTrue(result == OK, result);
+
+			/* --- Check required signatures --- */
+			final var treasuryId = Id.fromGrpcAccount(tokenCreateOp.getTreasury());
+			final var treasuryHasSigned = validateKey(frame, treasuryId.asEvmAddress(), sigsVerifier::hasActiveKey);
+			validateTrue(treasuryHasSigned, INVALID_SIGNATURE);
+			tokenCreateOp.getAdminKey().ifPresent(key -> validateTrue(validateAdminKey(frame, key), INVALID_SIGNATURE));
+
+			/* --- Build the necessary infrastructure to execute the transaction --- */
+			final var scopedAccountStore = createAccountStore();
+			final var scopedTokenStore = createTokenStore(scopedAccountStore, sideEffectsTracker);
+			final var tokenCreateLogic = createLogicFactory.newTokenCreateLogic(scopedAccountStore, scopedTokenStore,
+					dynamicProperties, sigImpactHistorian, sideEffectsTracker, entityIdSource, validator);
+
+			/* --- Execute the transaction and capture its results --- */
+			tokenCreateLogic.create(creationTime.getEpochSecond(), EntityIdUtils.accountIdFromEvmAddress(senderAddress),
+					transactionBody.getTokenCreation());
+		}
+
+		@Override
+		public void handleSentHbars(final MessageFrame frame) {
+			final var timestampSeconds = frame.getBlockValues().getTimestamp();
+			final var timestamp = Timestamp.newBuilder().setSeconds(timestampSeconds).build();
+			final var gasPriceInTinybars = feeCalculator.get()
+					.estimatedGasPriceInTinybars(ContractCall, timestamp);
+			final var calculatedFeeInTinybars = gasFeeInTinybars(
+					transactionBody.setTransactionID(TransactionID.newBuilder().setTransactionValidStart(
+							timestamp).build()), Instant.ofEpochSecond(timestampSeconds));
+
+			final var tinybarsRequirement = calculatedFeeInTinybars + (calculatedFeeInTinybars / 5)
+					- precompile.getMinimumFeeInTinybars(timestamp) * gasPriceInTinybars;
+
+			validateTrue(messageFrame.getValue().greaterOrEqualThan(Wei.of(tinybarsRequirement)), INSUFFICIENT_TX_FEE);
```

Comment:
Why can't we use the `frame` parameter from like 994?

Reply:
Same comment as above --- this one is even more embarrassing 😞